MAINT/BUG: Remove special-casing for 0d arrays, now that indexing with a single boolean is ok

eric-wieser opened the pull request:
Also fix the test added in gh-4792, which didn't make sense, but passed anyway

eric-wieser committed:
MAINT/BUG: Remove special-casing for 0d arrays, now that indexing with a single boolean is ok

Also fix the test added in gh-4792, which didn't make sense, but passed anyway

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1323,21 +1323,15 @@ def piecewise(x, condlist, funclist, *args, **kw):
-    if (isscalar(condlist) or not (isinstance(condlist[0], list) or
-                                   isinstance(condlist[0], ndarray))):
-        if not isscalar(condlist) and x.size == 1 and x.ndim == 0:
-            condlist = [[c] for c in condlist]
```
Essentially, this line is replaced with `pass`, and then the nested `if`s are collapsed

eric-wieser commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -2538,7 +2538,7 @@ def test_0d(self):
         assert_(y == 0)
 
         x = 5
-        y = piecewise(x, [[True], [False]], [1, 0])
```
This didn't make any sense, and should have failed. The conditions should be the same shape as the `x`

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1323,21 +1323,15 @@ def piecewise(x, condlist, funclist, *args, **kw):
-    zerod = False
-    if x.ndim == 0:
-        x = x[None]
-        zerod = True
```
I think this worked around `arr_0d[bool_0d]` not being supported in the past

eric-wieser committed:
DOC: piecewise callables take 1d arrays

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -1323,21 +1323,15 @@ def piecewise(x, condlist, funclist, *args, **kw):
-            condlist = [condlist]
+
+    # undocumented: single condition is promoted to a list of one condition
+    if np.ndim(condlist) == 0 or (
```
This took me very long to parse; could you add a comment for what the second part of the if statement is for? It might have helped me a little to have seen a reversed order `x.ndim >0 and not isinstance(...)`

eric-wieser replied:
So, part of the problem is that this statement is to hit an arbitrary selection of cases where the user maybe forgot to pass a list. It performs the following mappings

* `piecewise(x_1d, cond_1d, ...)` -> `piecewise(x_1d, [cond_1d])`
* `piecewise(x_0d, cond_0d, ...)` -> `piecewise(x_1d, [cond_0d])`  (after the second commit)

But doesn't do anything to

*  `piecewise(x_2d, cond_2d, ...)`-
*  `piecewise(x_3d, cond_3d, ...)`

Which it interprets as broadcasting an `n-1`d condition against an `nd` array.

This is super inconsistent - perhaps I should leave the `isscalar`, and just deprecate the entire code path?

mhvk reviewed (approved):
Only a nitpick about a comment.

eric-wieser committed:
TST: Add test for 0d conditions in np.piecewise

eric-wieser commented:
I've removed the `isscalar` &rarr; `ndim` change, since there's no reason to increase the scope of an undocumented feature if the better option is to deprecate it.

eric-wieser committed:
BUG: Throw an error if too many functions are given to piecewise

Especially necessary given the strange heuristics that decay the number of conditions to 1

mhvk commented on numpy/lib/function_base.py:
```diff
@@ -1323,25 +1323,24 @@ def piecewise(x, condlist, funclist, *args, **kw):
+    if isscalar(condlist) or (
+            not isinstance(condlist[0], (list, ndarray)) and x.ndim != 0):
+        condlist = [condlist]
+
```
So, was the idea to deprecate this? An alternative might be to just add `ndmin=1` in the conversion to boolean array below.

eric-wieser replied:
I'd like to deprecate this in future, but this patch is orthogonal to that

eric-wieser replied:
`ndim=1` won't help - this is intended to also promote 1d arrays to a list of length 1 containing a single 1d array.

mhvk commented:
OK, makes sense to postpone `isscalar` to another PR. So, things look OK except that the test failures are real.

eric-wieser commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -2556,6 +2561,17 @@ def test_0d_comparison(self):
+        assert_raises_regex(ValueError, '2 or 3 functions are expected',
+            piecewise, x, x, [x <= 3, x > 3], [1])
+        assert_raises_regex(ValueError, '2 or 3 functions are expected',
+            piecewise, x, x, [x <= 3, x > 3], [1, 1, 1, 1])
```
Typos here - should be `x` just once. Alarming that only some platforms are failing...

charris commented:
Thanks Eric.
